tsfc/coffee_mode.py

```diff
@@ -0,0 +1,338 @@
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+    return [optimise(node, quadrature_multiindex, argument_multiindices) for node in expressions]
+
+
+def index_extent(factor, argument_indices):
+    """Compute the product of the extents of argument indices of a GEM expression
+
+    :arg factor: GEM expression
+    :arg argument_indices: set of argument indices
+
+    :returns: product of extents of argument indices
+    """
+    return numpy.product([i.extent for i in set(factor.free_indices).intersection(argument_indices)])
+
+
+def unique_sum_indices(monomial_sum):
+    """Create a generator of unique sum indices of monomials in a monomial sum.
+
+    :arg monomial_sum: :class:`MonomialSum` object
+
+    :returns: a generator of unique sum indices
+    """
+    seen = set()
+    for monomial in monomial_sum:
+        fs = frozenset(monomial.sum_indices)
+        if fs not in seen:
+            seen.add(fs)
+            yield monomial.sum_indices
+
+
+def monomial_sum_to_expression(monomial_sum):
+    """Convert a monomial sum to a GEM expression. Uses associate_product() and
+    associate_sum() to promote hoisting in the subsequent code generation.
+
+    :arg monomial_sum: :class:`MonomialSum` object
+
+    :returns: GEM expression
+    """
+    indexsums = []  # The result is summation of indexsums
+    sum_indices_set_map = {}  # fronzenset(sum_indices) -> sum_indices
+    monomial_groups = OrderedDict()  # frozonset(sum_indices) -> [(atomics, rest)]
+    # Group monomials according to their sum indices
+    for monomial in monomial_sum:
+        if not monomial.sum_indices:
+            # IndexSum(reduce(Product, atomics, rest), sum_indices)
+            product, _ = associate_product(monomial.atomics + (monomial.rest,))
+            indexsums.append(product)
+        else:
+            fs = frozenset(monomial.sum_indices)
+            sum_indices_set_map.setdefault(fs, monomial.sum_indices)
+            monomial_groups.setdefault(fs, []).append((monomial.atomics, monomial.rest))
+
+    # Create IndexSum's from each monomial group
+    for sum_indices_set, list_atomics_rest in iteritems(monomial_groups):
+        sum_indices = sum_indices_set_map[sum_indices_set]
+        all_atomics, all_rest = zip(*list_atomics_rest)
+        if len(all_atomics) == 1:
+            # Just one term, add to indexsums directly
+            atomics, = all_atomics
+            rest, = all_rest
+            product, _ = associate_product(atomics + (rest,))
+            indexsums.append(IndexSum(product, sum_indices))
+        else:
+            # Create one product for each monomial
+            products = [associate_product(atomics + (rest,))[0] for atomics, rest in zip(all_atomics, all_rest)]
+            indexsums.append(IndexSum(associate_sum(products)[0], sum_indices))
+
+    return associate_sum(indexsums)[0]
```

Comment:
These `[0]`s are terribly unintuitive. But I recognise that sometimes you just want to build an optimised product/sum tree, and don't care about actual operation count. I suggest you make the return value of `associate_sum`/`associate_product` a named tuple instead of an ordinary tuple, for example,
```python
ExpressionAndFlopCount = namedtuple('ExpressionAndFlopCount', ['expression', 'flops'])
```
and then you can use the more descriptive `.expression` instead of `[0]`.